presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,191 @@
+    @GuardedBy("this")
+    private final LeakedQueries leakedQueries;
+
+    // We don't rely on the size of the weak leakedQueries map for the number of leaked queries as we want an exact counter.
+    private final AtomicInteger numberOfLeakedQueries = new AtomicInteger();
+
+    static ClusterMemoryLeakDetector createMemoryLeakDetector()
+    {
+        return new ClusterMemoryLeakDetector(new DefaultLeakedQueries());
+    }
+
+    @VisibleForTesting
+    static ClusterMemoryLeakDetector createTestMemoryLeakDetector()
+    {
+        return new ClusterMemoryLeakDetector(new TestLeakedQueries());
+    }
+
+    private ClusterMemoryLeakDetector(LeakedQueries leakedQueries)
+    {
+        this.leakedQueries = requireNonNull(leakedQueries, "leakedQueries is null");
+    }
+
+    /**
+     * @param allQueries All queries that the coordinator knows about.
+     * @param queryMemoryReservations The memory reservations of queries in the GENERAL cluster memory pool.
+     */
+    void checkForMemoryLeaks(List<QueryInfo> allQueries, Map<QueryId, Long> queryMemoryReservations)
+    {
+        requireNonNull(allQueries);
+        requireNonNull(queryMemoryReservations);
+
+        Map<QueryId, Long> leakedQueryReservations = queryMemoryReservations.entrySet()
+                .stream()
+                .filter(Objects::nonNull)
+                .filter(entry -> entry.getValue() > 0)
+                .filter(entry -> shouldClaimQueryAsLeaked(allQueries, entry.getKey()))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+
+        if (!leakedQueryReservations.isEmpty()) {
+            log.warn("Memory leak detected. The following queries are already finished, " +
+                    "but they have memory reservations on some worker node(s): %s", leakedQueryReservations);
+        }
+
+        // keep numberOfLeakedQueries and leakedQueries in sync
+        synchronized (this) {
+            numberOfLeakedQueries.set(leakedQueryReservations.size());
+            leakedQueries.update(leakedQueryReservations.keySet());
+        }
+    }
+
+    private static boolean shouldClaimQueryAsLeaked(List<QueryInfo> allQueries, QueryId queryId)
+    {
+        boolean queryExists = false;
+        QueryState queryState = null;
+        DateTime queryEndTime = null;
+
+        for (QueryInfo queryInfo : allQueries) {
+            if (queryInfo.getQueryId().equals(queryId)) {
+                queryExists = true;
+                queryState = queryInfo.getState();
+                queryEndTime = queryInfo.getQueryStats().getEndTime();
+            }
+        }
+
+        if (!queryExists) {
+            return true;
+        }
+
+        if (queryState.equals(RUNNING) || queryEndTime == null) {
+            return false;
+        }
+
+        return secondsBetween(queryEndTime, now()).getSeconds() >= DEFAULT_LEAK_CLAIM_DELTA_SEC;
+    }
+
+    synchronized boolean isQueryLeaked(QueryId queryId)
```

Comment:
`wasQueryLeaked`? I understand this method inspects a snapshot of state, not the current state